### PR TITLE
[SYCL][Tools] sycl-trace: extract backend specific code to dynamically loaded libraries

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
Not all systems could have cuda driver installed. Previous implementation with statically linked cuda blocks sycl-trace from being run on systems w/o cuda. Move L0 specific and CUDA specific collectors to dynamic libraries loaded by request.

NOTE: sycl-trace originally is not adapted to any other OS but Linux. So no general classes and other unified stuff is implemented.

### CODE REVIEW

File: sycl/tools/sycl-trace/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,7 @@
 add_executable(sycl-trace main.cpp)
 
+#add_definitions(-O0 -gdwarf-4)
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/CMakeLists.txt:3]: remove

File: sycl/tools/sycl-trace/CMakeLists.txt [hunk 2]
```diff
@@ -68,7 +79,10 @@ if ("level_zero" IN_LIST SYCL_ENABLE_PLUGINS)
     ${LEVEL_ZERO_INCLUDE_DIR}/ze_api.h
     )
 
-  add_dependencies(sycl_pi_trace_collector ze-pretty-printers)
+  add_dependencies(ze_trace_collector ze-pretty-printers)
+  target_compile_definitions(sycl_pi_trace_collector PRIVATE SYCL_HAS_LEVEL_ZERO)
+  #target_compile_options(ze_trace_collector PUBLIC
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/CMakeLists.txt:84]: remove

File: sycl/tools/sycl-trace/CMakeLists.txt [hunk 3]
```diff
@@ -142,15 +156,21 @@ if(SYCL_BUILD_PI_CUDA)
...
 
-  add_dependencies(sycl_pi_trace_collector cuda-pretty-printers)
-
+  add_dependencies(cuda_trace_collector cuda-pretty-printers)
+  target_link_libraries(cuda_trace_collector PRIVATE xptifw)
+  target_compile_definitions(sycl_pi_trace_collector
+    PRIVATE
+      $<$<BOOL:${SYCL_BUILD_PI_CUDA}>:USE_PI_CUDA>
+  )
+  #target_compile_options(cuda_trace_collector PUBLIC
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/CMakeLists.txt:165]: remove

File: sycl/tools/sycl-trace/cuda_trace_collector.cpp [hunk 4]
```diff
@@ -45,14 +41,15 @@ static std::string getResult(CUresult Res) {
   return ResultStr;
 }
 
-XPTI_CALLBACK_API void cuCallback(uint16_t TraceType,
+extern "C" {
+__attribute__((__visibility__("default"))) void callback(uint16_t TraceType,
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/cuda_trace_collector.cpp:45]: add define and header file

File: sycl/tools/sycl-trace/collector.cpp [hunk 5]
```diff
@@ -8,21 +8,144 @@
...
 #include "xpti/xpti_trace_framework.h"
 
+#include <dlfcn.h>
+#include <iostream>
+#include <mutex>
+#include <string>
 #include <sycl/detail/spinlock.hpp>
 
 sycl::detail::SpinLock GlobalLock;
+bool HasZEPrinter;
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/collector.cpp:18]: initialize please

File: sycl/tools/sycl-trace/collector.cpp [hunk 6]
```diff
@@ -8,21 +8,144 @@
...
+
+  void *MHandle = NULL;
+
+  void *MInitPtr = NULL;
+  void *MFinishPtr = NULL;
+  void *MCallbackPtr = NULL;
+  void *MSetIndentationLevelPtr = NULL;
+} zeCollectorLibrary("libze_trace_collector.so"),
+    cudaCollectorLibrary("libcuda_trace_collector.so");
+
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/collector.cpp:129]: add comment why we need this routing function

File: sycl/tools/sycl-trace/collector.cpp [hunk 7]
```diff
@@ -96,12 +214,14 @@ XPTI_CALLBACK_API void xptiTraceFinish(const char *StreamName) {
   else if (std::string_view(StreamName) ==
                "sycl.experimental.level_zero.debug" &&
            std::getenv("SYCL_TRACE_ZE_ENABLE"))
-    zePrintersFinish();
+    zeCollectorLibrary.finishPrinters();
+  zeCollectorLibrary.clear();
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/collector.cpp:218]: add {} and same indentation

File: sycl/tools/sycl-trace/collector.cpp [hunk 8]
```diff
@@ -96,12 +214,14 @@ XPTI_CALLBACK_API void xptiTraceFinish(const char *StreamName) {
...
 #endif
 #ifdef USE_PI_CUDA
   else if (std::string_view(StreamName) == "sycl.experimental.cuda.debug" &&
            std::getenv("SYCL_TRACE_CU_ENABLE"))
-    cuPrintersFinish();
+      cudaCollectorLibrary.finishPrinters();
+  cudaCollectorLibrary.clear();
```
Comment (KseniyaTikhomirova) [inline sycl/tools/sycl-trace/collector.cpp:224]: add {} and same indentation

File: sycl/tools/sycl-trace/collector.cpp [hunk 9]
```diff
@@ -8,21 +8,146 @@
...
+  if (0 == RetCode) {
+    // This actually indicates an error
+    return "";
+  }
+
+  auto Path = std::string(Info.dli_fname);
+  auto LastSlashPos = Path.find_last_of('/');
+
+  return Path.substr(0, LastSlashPos);
+}
```
Comment (steffenlarsen) [inline sycl/tools/sycl-trace/collector.cpp:34]: Is this meant to work on Windows? If so, does it?

File: sycl/tools/sycl-trace/collector.cpp [hunk 10]
```diff
@@ -8,21 +8,146 @@
...
+      : MLibraryName(LibraryName){};
+  ~CollectorLibraryWrapper() { clear(); };
+
+  const std::string InitFuncName = "init";
+  const std::string FinishFuncName = "finish";
+  const std::string CallbackFuncName = "callback";
+  const std::string IndentFuncName = "setIndentationLevel";
+
+  bool initPrinters() {
+    const std::string CollectorDir = getCurrentDSODir() + "/" + MLibraryName;
```
Comment (steffenlarsen) [inline sycl/tools/sycl-trace/collector.cpp:55]: Based on the above, `getCurrentDSODir()` can return the empty string in certain error cases. If this happens, this directory path would be `/some_library_name` which could be a problem. Instead, the value returned by `getCurrentDSODir()` should be checked first.

File: sycl/tools/sycl-trace/cuda_trace_collector.cpp [hunk 11]
```diff
@@ -119,4 +116,9 @@ void cuPrintersInit() {
 }
 
 // For unification purpose
-void cuPrintersFinish() {}
+COLLECTOR_EXPORT_API void finish() {}
+
+COLLECTOR_EXPORT_API void setIndentationLevel(int NewLevel) {
+  IndentationLevel = NewLevel;
+}
+}
```
Comment (steffenlarsen) [inline sycl/tools/sycl-trace/cuda_trace_collector.cpp:124]: Nit;
```suggestion
}

```

File: sycl/tools/sycl-trace/ze_trace_collector.cpp [hunk 12]
```diff
@@ -320,4 +319,9 @@ void zePrintersInit() {
 }
 
 // For unification purpose
-void zePrintersFinish() {}
+COLLECTOR_EXPORT_API void finish() {}
+
+COLLECTOR_EXPORT_API void setIndentationLevel(int NewLevel) {
+  IndentationLevel = NewLevel;
+}
+}
```
Comment (steffenlarsen) [inline sycl/tools/sycl-trace/ze_trace_collector.cpp:327]: Nit;
```suggestion
}

```

File: sycl/tools/sycl-trace/common.h
Comment (steffenlarsen) [inline sycl/tools/sycl-trace/common.h:1]: Is this meant to be a new file?

File: sycl/tools/sycl-trace/collector.cpp (hunk 9 above)
Comment (npmiller) [reply sycl/tools/sycl-trace/collector.cpp:34]: We should probably try to re-use `sycl/detail/os_util.cpp`  for this

File: sycl/tools/sycl-trace/collector.cpp [hunk 13]
```diff
@@ -8,21 +8,146 @@
...
+    MHandle = dlopen(CollectorDir.c_str(), RTLD_LAZY);
```
Comment (npmiller) [inline sycl/tools/sycl-trace/collector.cpp:56]: Does this need to work on Windows as well? Should we have Windows version with `LoadLibraryA` and so on

File: sycl/tools/sycl-trace/collector.cpp [hunk 14]
```diff
@@ -8,21 +8,146 @@
...
+private:
+  std::string MLibraryName;
+  int MIndentationLevel = 0;
+
+  void *MHandle = NULL;
+
+  void *MInitPtr = NULL;
+  void *MFinishPtr = NULL;
+  void *MCallbackPtr = NULL;
+  void *MSetIndentationLevelPtr = NULL;
```
Comment (npmiller) [inline sycl/tools/sycl-trace/collector.cpp:126]: nit: should we use `nullptr` instead of `NULL`?

File: sycl/tools/sycl-trace/collector.cpp (hunk 9 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/collector.cpp:34]: no, it is not intended to work on Windows. sycl-trace originally is not adapted for Windows so I didn't invest into it

File: sycl/tools/sycl-trace/collector.cpp (hunk 13 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/collector.cpp:56]: NOTE: sycl-trace originally is not adapted to any other OS but Linux. So no general classes and other unified stuff is implemented.

File: sycl/tools/sycl-trace/collector.cpp (hunk 9 above)
Comment (steffenlarsen) [reply sycl/tools/sycl-trace/collector.cpp:34]: Fair! Would @npmiller's suggestion to use `sycl/detail/os_util.cpp` work though? Would mean a smaller maintenance surface and less work to make it work on Windows if need be in the future.

Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/collector.cpp:34]: it is internal file of sycl RT. Not sure if referencing it in separated tool is a good solution. But if we are talking about copy of file - no benefits for now.

Comment (steffenlarsen) [reply sycl/tools/sycl-trace/collector.cpp:34]: I was hoping maybe the tracer was able to share some files, but if that isn't the case then the current solution is probably better.

File: sycl/tools/sycl-trace/collector.cpp (hunk 14 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/collector.cpp:126]: "If handle does not refer to a valid object opened by [dlopen()](https://pubs.opengroup.org/onlinepubs/009696899/functions/dlopen.html), or if the named symbol cannot be found within any of the objects associated with handle, dlsym() shall return NULL."
I assume there should be no difference if I change it to nullptr. I am ok doing that. Although I tried to work in terms of used API.

File: sycl/tools/sycl-trace/common.h
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/common.h:1]: removed, thanks

File: sycl/tools/sycl-trace/ze_trace_collector.cpp (hunk 12 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/ze_trace_collector.cpp]: done

File: sycl/tools/sycl-trace/cuda_trace_collector.cpp (hunk 11 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/cuda_trace_collector.cpp]: done

File: sycl/tools/sycl-trace/collector.cpp (hunk 10 above)
Comment (KseniyaTikhomirova) [reply sycl/tools/sycl-trace/collector.cpp]: done, good catch, thanks

**Review by steffenlarsen — APPROVED** [review]
LGTM! 🚀

**Review by npmiller — APPROVED** [review]
LGTM, I've also tested on a CUDA machine and the tracing works fine with this change

**Comment by KseniyaTikhomirova** [comment]
ready for merge

**Pull request merged**